guardian/grid
# Addition of announcement/notification banner

## What does this change do?
This feature adds in a new React component to provide a notification/announcement banner that can display a variety of messages to the user. Initial use case will be to announce the release of new features but it has been designed so that in time it can be used for all user notifications - warnings, error messages, information etc.

The new features are announced by adding items to the announcements array in the application.conf

The addition of patterns to introduce other messages/notifications has been left for a future PR.

If no announcements are included in the application.conf the functionality will have no impact on the application.

The application polls the server one every minute to check for new notifications

The format of the announcement objects is;

[ (array)
    { (json object)
      **announceId**: (string) the unique id of the announcement - should be unique among all active announcements
      **description**: (string) the main text to display in the notification notification_banner
      **endDate**: (string, optional, format="yyyy-mm-dd") the date beyond which the announcement should not be seen, if not present set as today + 1 year
      **url**: (string, optional) a link to a page/document providing further details regarding announcement
      **urlText**: (string, optional) text to be included in a-tag hyperlink (will revert to default if not present)
      **category**: (string) the type of announcement - will control styling and display, Enum=_announcement, information, warning, error, success_
      **lifespan**: (string) the lifecycle behaviour Enum=_transient_ (message disappears on any click etc),
                                                      _session_ (message must be acknowledged but action NOT stored in client cookie - used for current session messages)
                                                      _persistent_ (message must be acknowledged and action stored in client cookie - used for long running announcements)
   },
   ...
 ]

## How should a reviewer test this change?

Add some example notifications to the application.conf and ensure that they displaty correctly and follow the correct lifetime rules (see lifespan options above)

A set of examples could be;

announcements = [
  {
    announceId: "notification_banner",
    description: "A New Feature Notification Banner",
    endDate: "2025-03-31",
    url: "https://runacharainn.com",
    urlText: "Runach Arainn Glamping",
    category: "announcement",
    lifespan: "persistent"
  },
  {
    announceId: "sort_control",
    description: "Sort Control",
    endDate: "2025-04-01",
    url: "https://www.bbc.co.uk",
    category: "warning",
    lifespan: "transient"
  },
  {
      announceId: "success_ex",
      description: "This is an example of a success message",
      endDate: "2025-04-01",
      category: "success",
      lifespan: "transient"
  },
  {
      announceId: "error_msg",
      description: "This is an example of a error message. Please click to acknowledge.",
      endDate: "2025-04-01",
      category: "error",
      lifespan: "session"
  },
  {
      announceId: "old_message",
      description: "This is an old message",
      endDate: "2024-01-01",
      category: "success",
      lifespan: "persistent"
  }
]

An example of how these notifications should look is;

![Screenshot 2024-03-19 at 09 40 38](https://github.com/guardian/grid/assets/128470622/6962353c-99ec-4edf-bee0-ad7863e5739c)

## Tested? Documented?
